### ISSUE TITLE
Course Restore with Quickmail History Fails

### ISSUE THREAD

**Issue by bfriesenvcc** [lsuits/block_quickmail#179]
I get the following error while trying to restore a course WITH Quickmail history set to "yes" (in courses with a history to restore). I'm using Moodle version: 2.6.10 (2013111810) and Quickmail version 2014042915, but the problem existed with 2014022713, as well. 

error/unknown_context_mapping
More information about this error
Debug info: 
Error code: unknown_context_mapping
$a contents: error
Stack trace:
• line 887 of /backup/util/dbops/restore_dbops.class.php: restore_dbops_exception thrown
• line 75 of /blocks/quickmail/backup/moodle2/restore_quickmail_stepslib.php: call to restore_dbops::send_files_to_pool()
• line 41 of /blocks/quickmail/backup/moodle2/restore_quickmail_stepslib.php: call to restore_quickmail_log_structure_step->process_log()
• line 137 of /backup/util/plan/restore_structure_step.class.php: call to restore_quickmail_log_structure_step->process_block()
• line 103 of /backup/util/helper/restore_structure_parser_processor.class.php: call to restore_structure_step->process()
• line 123 of /backup/util/xml/parser/processors/grouped_parser_processor.class.php: call to restore_structure_parser_processor->dispatch_chunk()
• line 204 of /backup/util/xml/parser/progressive_parser.class.php: call to grouped_parser_processor->after_path()
• line 284 of /backup/util/xml/parser/progressive_parser.class.php: call to progressive_parser->inform_end()
• line ? of unknownfile: call to progressive_parser->end_tag()
• line 175 of /backup/util/xml/parser/progressive_parser.class.php: call to xml_parse()
• line 154 of /backup/util/xml/parser/progressive_parser.class.php: call to progressive_parser->parse()
• line 110 of /backup/util/plan/restore_structure_step.class.php: call to progressive_parser->process()
• line 181 of /backup/util/plan/base_task.class.php: call to restore_structure_step->execute()
• line 177 of /backup/util/plan/base_plan.class.php: call to base_task->execute()
• line 167 of /backup/util/plan/restore_plan.class.php: call to base_plan->execute()
• line 333 of /backup/controller/restore_controller.class.php: call to restore_plan->execute()
• line 184 of /backup/util/ui/restore_ui.class.php: call to restore_controller->execute_plan()
• line 99 of /backup/restore.php: call to restore_ui->execute()

**Comment by rrusso** [comment]
Dave, is this included in you're latest round of fixes?

**Comment by elliott2081** [comment]
Hello @bfriesenvcc I wrote a comment to you on moodle.org, but I guess it didn't actually get posted. Sorry. The error you're experiencing is from a bug around that time. I would highly recommend updating to the latest version. It should work fine in 2.6. If it doesn't let me know asap, thanks!
Dave

**Comment by elliott2081** [comment]
I believe the problem she is facing originated and was repaired around this commit last year. 

https://github.com/lsuits/quickmail/commit/419888afa4938dbb0f940514c8224de660a21b59

**Comment by bfriesenvcc** [comment]
Sorry for the delay in response, guys.  

I am using the latest version of Quickmail, for Moodle 2.6 -- 2014042914 (I made a typo, last time and put 2014042915).   The error still shows.

**Comment by reccoller** [comment]
I am having the same issue, we went from Moodle 2.5.5 to 2.7.8 (20150511)

![quickmail-error](https://cloud.githubusercontent.com/assets/12550543/7758169/a52b177a-ffd5-11e4-88ea-f30c00cb658e.png)

**Comment by elliott2081** [comment]
@bfriesenvcc and @reccoller thanks. looking into it. What are you trying to do? If you try to restore settings from an earlier version to a later version, that won't work, because the older version didn't have that as an option so there won't be settings to restore when you get to that step.  
I will be testing this a little bit. 
If you can, could you email me some screenshots showing the backup and restore process as detailed as possible. The error log doesn't mean much to me. It lets me know where it crashed in the code, but doesn't provide many clues as to in what context or scenario. 
DB info would also be good, are you using MySQL? Oracle? SQLServer etc..

**Comment by reccoller** [comment]
We are using MySQL 2012

We have updated our Moodle server between semesters. (2.5.5  2.7.8) I have some instructors that would like to view some of the emails that were sent through the Quickmail plugin block. When I try to restore the whole full back up with users, this is the debug error I received. If I uncheck the box for Blocks on the restore process I can restore the course with ease. We use the automatic built in Moodle back up cron job everynight. For a full back up of everything in the course. 

We use all the default settings for the on the restore to restore everything...(the logs are not coming in as well - but that is a separate issue)

**Comment by hrynkiw** [comment]
We've been getting this error "unknown_context_mapping" as well when restoring course archive files
- generated by the automated Moodle backup process, with users
- created under Moodle v2.6, restored to Moodle v2.8
- I don't know what version of Quickmail we had under Moodle v2.6. Under v2.8, we've got Quickmail 2014022713, v1.3.2
- mySQL
- If I choose to include Quickmail in the backup, the restore fails. If I exclude Quickmail from the backup, the restore succeeds. Thank-you reccoller, for the tip re: de-selecting Blocks during the restore process as a work-around. We often don't have access to the original course to re-do the backup.

Luckily enough, none of our teachers have needed the Quickmail content restored so we were able to get around the error by excluding it from the restore. But I am not looking forward to the day when the Quickmail content restore is requested.

**Comment by elliott2081** [comment]
@bfriesenvcc  In your original message you said 
"(in courses with a history to restore)."
  Are you 100% sure they had a history to restore? I was looking through the commit and noticed I left a note for myself, which basically says that shows up as an option even when it shouldn't, I will be fixing that this weekend for sure.  @hrynkiw @reccoller kind of the same question, when youre doing auto backups are they for sure checking the include logs checkbox? By default I believe that's turned off, so if there are no logs and you say to restore logs, it will error. 
Even if everything is setup correctly, I think fixing that issue might fix a bug or two, that might be erroring out for no good reason. I'll hopefully be posting a message here Monday. :+1:

**Comment by elliott2081** [comment]
oops by 'logs' I mean quickmail history.

**Comment by hrynkiw** [comment]
A screenshot of our automated backup settings.
![kwantlen-automatedbackup-screenshot](https://cloud.githubusercontent.com/assets/4992018/7816650/6a0c06c4-0384-11e5-8d0a-37f091fa3994.jpg)

If I have time in the next couple of days, I'll also post a screenshot of the restore settings which generate the error message.

Hrynkiw
Kwantlen Polytechnic

**Comment by elliott2081** [comment]
Thanks! I am having a hard time reproducing the error, but am still looking into it. If someone finds the issue please let us know.

**Comment by bfriesenvcc** [comment]
@elliott2081 , sorry for the late response.  Yes, I'm 100% sure.  I just tested it again, to be sure.

**Comment by bfriesenvcc** [comment]
I'm checking back on this, as we have now upgraded to Moodle 2.8.7, Quickmail version 1.5.0 (2014042914) and we are still experiencing the problem. 

All I did was:
1) Added Quickmail to one of my test shells
2) Sent a message to my Student self
3) Made a quick full backup of the course shell
4) Downloaded the backup file
5) Attempted to restore the downloaded file, including all users, and with all default settings (Restore QM History: Yes; Overwrite QM History: No) to a new shell, changing only the name.

After SEVERAL minutes, I got the following error:
![quickmailerror](https://cloud.githubusercontent.com/assets/8193384/9231072/f23e8af6-40da-11e5-977f-a87165c1132b.PNG)

**Comment by bfriesenvcc** [comment]
FYI, I have just repeated the exercise, except this time I tried to restore the course into itself, as this would more closely resemble a real-life "disaster recovery" of a course shell. Same wait, same error message.

**Comment by bfriesenvcc** [comment]
I'm hoping someone is still looking into this, as the issue has come up, again, today.  Has anyone else been able to reproduce this?

**Comment by mchaney2010** [comment]
Hi, we are using Moodle 2.7.7 with quickmail version 2014042914 and we too are having this exact same issue. 

I have done some digging and it looks as if it all stems from the same problem, 
Line 35 of restore_quickmail_stepslib.php in the function process_block($data) makes a reference to '$current'. But $current has never been defined anywhere. This throws a warning as it tries to get a property of a non-object. $current is used in the paramaters (as newitemid) of a query which it uses to try and work out the $id of the backup_ids_temp. This id (which it can never find as the newitemid in the params wasn't set) is then passed to process_log function as '$oldctx'. The process_log function uses this $oldctx id in its send_files_to_pool function. This is where it throws the invalid_context_mapping because the $oldtx is null. Oh and it will then fail later on, once it gets passed that point, because the next query uses the 'context' variable passed down, which was the id from current, but as current is not set, then there is no id to use.

Hard to follow unless you look at the code, but hopefully it can help, basically, line 35, $current is not defined. 

At this stage I do not know what $current should be?

Thanks

Mark

**Comment by bfriesenvcc** [comment]
Thanks for your reply, Mark.  At least it's comforting to know I'm not alone.  :-)

**Comment by mchaney2010** [comment]
Dear all, 

I can confirm that the issue I explained above is indeed the issue with the code. 

To fix it: 
FIle: blocks\quickmail\backup\moodle2\restore_quickmail_stepslib.php
Line 35, where it says 'newitemid' => $current->id
Should be 'newitemid' => $context->id

Line 42 (ish), where it says $this->process_log($log, $id, $current);
Should be $this->process_log($log, $id, $context);

Or you could rename line 30, $context = context_course::instance($this->get_courseid()); to $current = context_course::instance($this->get_courseid());

But to me $context has more sense that $current.

**Comment by bfriesenvcc** [comment]
Thanks for providing that, Mark. Hopefully, that will get included in the next release.

**Comment by elliott2081** [comment]
Thanks Mark. I have put those corrections to the MOODLE_STABLE_30 branch.

**Comment by markchaneyuk** [comment]
Please can you update your version in the Moodle.org plugins database as the version that is there still contains this error.

**Comment by rrusso** [comment]
We will be updating the moodle plugins repo shortly. @chadmazilly has now been assigned Quickmail and will be keeping tabs on things while rewriting the block for 3.2, 3.3, and onward.
